keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,436 @@
+source IP is preserved, and I _do_ care about maximizing reliability
+when pods restart or move between nodes (which can cause hiccups with
+`externalTrafficPolicy: Local` in some cases).
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
+(and is not allowed to fall back when there are no local endpoints),
+and `externalTrafficPolicy: Local` requires source IP preservation
+(which is generally not possible when sending to a non-local
+endpoint).
+
+### Choosing Services to Apply Node-Level Topology To
+
+```
+<<[UNRESOLVED]>>
+
+Pick an approach.
+
+The Automatic approach seems simplest (and would have the effect of
+immediately making DNS faster in most(?) clusters). But maybe it would
+do bad things in some circumstances?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Approach 1: Manual
+
+One approach would be to allow users to explicitly tag services where
+they want this behavior.
+
+This could be implemented via the existing
+`"service.kubernetes.io/topology-aware-hints"` annotation perhaps.
+Instead of setting it to `Auto`, the user could set it to `Node` to
+indicate node-level topology rather than zone-level.
+
+(This would imply that a service could not use both node-level
+topology and zone-level topology. Another possibility would be to have
+a new annotation for node-level topology.)
+
+#### Approach 2: Automatic
+
+Alternatively, the EndpointSlice controller could attempt to determine
+automatically which services would benefit from node-level topology.
+For example, the EndpointSlice controller could look for services
+where:
+
+  - The service has an endpoint on every node, or at least "almost
+    every" node. (eg, no more than N or N% of nodes are missing an
+    endpoint).
+
+  - The pods that make up the service endpoints all have an
+    `OwnerReference` pointing to the same `DaemonSet`.
```

Comment:
This seems pretty restrictive but could be a reasonable starting point. May have messy transition states if a DaemonSet is replaced.

Reply:
It's not really that messy; when the first few nodes lose their endpoints for the old service, they would automatically fall back to using Cluster semantics. After enough nodes lost endpoints, the EndpointSlice controller would turn off node-level topology for the service and all the remaining nodes would flip to Cluster semantics. There would be less total churn than there would be without node-level topology, since the first few endpoint deletions would be ignored by most of the nodes.

Reply:
Why do we need this criterion at all?

Reply:
See [this comment](https://github.com/kubernetes/enhancements/pull/3293/files#r865307808) in a collapsed thread. (TL;DR: "service that is intentionally deployed to every node" is semantically different from "service which coincidentally happens to be deployed to every node" in relevant ways.)

Reply:
Here's something I have seen:

Cluster has 3 node-pools: small, medium, large, with different sized machines.  They run 3 different daemonsets with node selectors for each pool, and different resource requests on each.

"Prefer local" would be ideal but fails this criterion.